File: WebClient SDK/src/default/ui/cardFiles/customCardFilesDataProvider.ts
import { IStorage } from 'tessa/platform/storage';
import { SchemeType } from 'tessa/scheme/schemeType';
import {
  FileListViewModel,
  FileViewModel,
  ViewControlDataProviderResponse
} from 'tessa/ui/cards/controls';
import { ViewMetadataSealed } from 'tessa/views/metadata';
import { ExternalFile } from '../../externalFiles/externalFile';
import { CardFilesDataProvider } from './cardFilesDataProvider';

/**
 * Customized provider for files in view with addition of Creation date column and corresponding data.
 */
export class CustomCardFilesDataProvider extends CardFilesDataProvider {
  constructor(readonly viewMetadata: ViewMetadataSealed, readonly fileControl: FileListViewModel) {
    super(viewMetadata, fileControl);
  }

  /**
   * Fill in data columns descriptions in data response.
   * @param result response with columns filled in
   */
  protected addFieldDescriptions(result: ViewControlDataProviderResponse): void {
    super.addFieldDescriptions(result);
    result.columns.push(['Date', SchemeType.DateTime]);
    result.columns.push(['Description', SchemeType.NullableString]);
  }

  /**
   * Fill in data row based on given file view model.
   * @param file file view model
   * @returns one row in data response.
   */
  protected mapFileToRow(file: FileViewModel): IStorage {
    const row = super.mapFileToRow(file);
    row['Date'] = file.model.created;
    let description: string | Blob | null = null;
    const externalFile = file.model as ExternalFile;
    if (externalFile) {
      description = externalFile.description;
    }
    row['Description'] = description;
    return row;
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/CardFiles/CustomCardFilesDataProvider.cs
#nullable enable

using System.Collections.Generic;
using Tessa.Extensions.Default.Client.ExternalFiles;
using Tessa.Scheme;
using Tessa.UI.Cards.Controls;
using Tessa.UI.Files;
using Tessa.Views.Metadata;

namespace Tessa.Extensions.Default.Client.UI.CardFiles
{
    /// <summary>
    /// Custom data provider for files in view extension.
    /// </summary>
    public sealed class CustomCardFilesDataProvider :
        CardFilesDataProvider
    {
        #region Constructor

        /// <summary>
        /// Create new instance of custom data provider.
        /// </summary>
        /// <param name="viewMetadata"><inheritdoc cref="IViewMetadata" path="/summary"/></param>
        /// <param name="fileControl"><inheritdoc cref="IFileControl" path="/summary"/></param>
        public CustomCardFilesDataProvider(IViewMetadata viewMetadata, IFileControl fileControl)
            : base(viewMetadata, fileControl)
        {
        }

        #endregion

        #region Base Overrides

        /// <inheritdoc/>
        protected override void AddFieldDescriptions(IGetDataResponse result)
        {
            base.AddFieldDescriptions(result);
            result.Columns.Add(new("Date", SchemeType.DateTime));
            result.Columns.Add(new("Description", SchemeType.NullableString));
        }

        /// <inheritdoc/>
        protected override Dictionary<string, object?> MapFileToRow(IFileViewModel file)
        {
            var row = base.MapFileToRow(file);
            row["Date"] = file.Model.Created;
            row["Description"] = (file.Model as ExternalFile)?.Description;
            return row;
        }

        #endregion
    }
}
